File: go/open_redirect/account_management.go
package openredirect

import (
	"net/http"
)

// VULN 1: Unvalidated `redirect_to` param used in redirect after password reset confirm - open redirect
func PasswordResetConfirmHandler(w http.ResponseWriter, r *http.Request) {
	token := r.FormValue("token")
	newPassword := r.FormValue("new_password")
	redirectTo := r.URL.Query().Get("redirect_to")
	if !validateResetToken(token) {
		http.Error(w, "Invalid or expired token", http.StatusBadRequest)
		return
	}
	_ = updatePassword(token, newPassword)
	if redirectTo == "" {
		redirectTo = "/login"
	}
	http.Redirect(w, r, redirectTo, http.StatusFound)
}

// VULN 2: Referer header used directly as logout redirect destination - open redirect
func LogoutHandler(w http.ResponseWriter, r *http.Request) {
	http.SetCookie(w, &http.Cookie{
		Name:   "session_token",
		Value:  "",
		MaxAge: -1,
	})
	referer := r.Header.Get("Referer")
	if referer == "" {
		referer = "/login"
	}
	http.Redirect(w, r, referer, http.StatusFound)
}

// VULN 3: Unvalidated `callback_url` used in redirect after social account linking - open redirect
func SocialLinkCallbackHandler(w http.ResponseWriter, r *http.Request) {
	provider := r.URL.Query().Get("provider")
	code := r.URL.Query().Get("code")
	callbackURL := r.URL.Query().Get("callback_url")
	if provider == "" || code == "" {
		http.Error(w, "Missing provider or code", http.StatusBadRequest)
		return
	}
	_ = linkSocialAccount(provider, code)
	if callbackURL == "" {
		callbackURL = "/account/settings"
	}
	http.Redirect(w, r, callbackURL, http.StatusFound)
}

func validateResetToken(token string) bool             { return true }
func updatePassword(token, newPassword string) error   { return nil }
func linkSocialAccount(provider, code string) error    { return nil }

File: dotnet/open_redirect/AccountManagement.cs
using Microsoft.AspNetCore.Mvc;

public class AccountManagementController : Controller
{
    // VULN 1: Unvalidated `redirect_to` query parameter passed directly to Redirect() after password reset confirm
    [HttpPost("reset-password-confirm")]
    public IActionResult ConfirmPasswordReset([FromForm] string token, [FromForm] string newPassword, [FromQuery] string redirect_to)
    {
        ResetPassword(token, newPassword);
        return Redirect(redirect_to ?? "/login");
    }

    // VULN 2: Referer header used directly as redirect target on logout - open redirect via crafted Referer
    [HttpPost("logout")]
    public IActionResult Logout()
    {
        InvalidateSession();
        string referer = Request.Headers["Referer"].ToString();
        string redirectTarget = string.IsNullOrEmpty(referer) ? "/" : referer;
        return Redirect(redirectTarget);
    }

    // VULN 3: Unvalidated `callback_url` parameter passed to Redirect() after social account linking
    [HttpGet("social-link-callback")]
    public IActionResult SocialAccountLinkCallback([FromQuery] string provider, [FromQuery] string callback_url)
    {
        LinkSocialAccount(provider);
        return Redirect(callback_url ?? "/account/settings");
    }

    private void ResetPassword(string token, string pass) { }
    private void InvalidateSession() { }
    private void LinkSocialAccount(string provider) { }
}
